File: static/js/charts/sankey.js
/**
 * D3 Sankey diagram implementation for customer flow visualization.
 * Used in Essay 1: How Customers Actually Become Loyal
 */

function renderSankey(container, spec) {
    const { data, config } = spec;
    const { nodes, links } = data;

    const width = config.width || 800;
    const height = config.height || 500;
    const nodeWidth = config.nodeWidth || 20;
    const nodePadding = config.nodePadding || 10;

    // Clear container
    d3.select(container).selectAll('*').remove();

    const svg = d3.select(container)
        .append('svg')
        .attr('width', '100%')
        .attr('height', height)
        .attr('viewBox', `0 0 ${width} ${height}`);

    // Create sankey generator
    const sankey = d3.sankey()
        .nodeWidth(nodeWidth)
        .nodePadding(nodePadding)
        .nodeAlign(d3.sankeyJustify)
        .extent([[20, 20], [width - 20, height - 20]]);

    // Generate sankey data
    const sankeyData = sankey({
        nodes: nodes.map(d => Object.assign({}, d)),
        links: links.map(d => Object.assign({}, d))
    });

    // Color scale
    const colors = config.colors || {
        'new': '#4299e1',
        'one_time': '#fc8181',
        'irregular': '#f6ad55',
        'regular': '#68d391',
        'long_cycle': '#b794f4'
    };

    // Draw links
    const link = svg.append('g')
        .attr('class', 'links')
        .attr('fill', 'none')
        .attr('stroke-opacity', config.linkOpacity || 0.5)
        .selectAll('path')
        .data(sankeyData.links)
        .join('path')
        .attr('class', 'sankey-link')
        .attr('d', d3.sankeyLinkHorizontal())
        .attr('stroke', d => colors[d.source.id] || '#a0aec0')
        .attr('stroke-width', d => Math.max(1, d.width))
        .on('mouseover', function(event, d) {
            d3.select(this)
                .attr('stroke-opacity', config.linkOpacityHover || 0.8);
            showTooltip(event, `${d.source.name} → ${d.target.name}: ${d.value.toLocaleString()}`);
        })
        .on('mouseout', function() {
            d3.select(this)
                .attr('stroke-opacity', config.linkOpacity || 0.5);
            hideTooltip();
        });

    // Draw nodes
    const node = svg.append('g')
        .attr('class', 'nodes')
        .selectAll('g')
        .data(sankeyData.nodes)
        .join('g')
        .attr('class', 'sankey-node');

    node.append('rect')
        .attr('x', d => d.x0)
        .attr('y', d => d.y0)
        .attr('height', d => d.y1 - d.y0)
        .attr('width', d => d.x1 - d.x0)
        .attr('fill', d => colors[d.id] || '#a0aec0')
        .attr('stroke', '#fff')
        .attr('stroke-width', 1)
        .attr('rx', 4);

    // Node labels
    node.append('text')
        .attr('x', d => d.x0 < width / 2 ? d.x1 + 6 : d.x0 - 6)
        .attr('y', d => (d.y1 + d.y0) / 2)
        .attr('dy', '0.35em')
        .attr('text-anchor', d => d.x0 < width / 2 ? 'start' : 'end')
        .attr('font-size', '12px')
        .attr('fill', '#4a5568')
        .text(d => `${d.name} (${d.value?.toLocaleString() || 0})`);

    // Animation helpers
    return {
        highlightNode: function(nodeId) {
            node.selectAll('rect')
                .attr('opacity', d => d.id === nodeId ? 1 : 0.3);
            link.attr('stroke-opacity', d =>
                d.source.id === nodeId || d.target.id === nodeId
                    ? 0.8 : 0.1
            );
        },
        highlightFlow: function(sourceId, targetId) {
            link.attr('stroke-opacity', d =>
                d.source.id === sourceId && d.target.id === targetId
                    ? 0.8 : 0.1
            );
        },
        showAll: function() {
            node.selectAll('rect').attr('opacity', 1);
            link.attr('stroke-opacity', config.linkOpacity || 0.5);
        }
    };
}

// Tooltip helpers
function showTooltip(event, content) {
    let tooltip = d3.select('.chart-tooltip');
    if (tooltip.empty()) {
        tooltip = d3.select('body')
            .append('div')
            .attr('class', 'chart-tooltip')
            .style('position', 'absolute')
            .style('background', '#1a202c')
            .style('color', 'white')
            .style('padding', '8px 12px')
            .style('border-radius', '6px')
            .style('font-size', '14px')
            .style('pointer-events', 'none')
            .style('z-index', '100');
    }
    tooltip
        .html(content)
        .style('left', (event.pageX + 10) + 'px')
        .style('top', (event.pageY - 10) + 'px')
        .style('opacity', 1);
}

function hideTooltip() {
    d3.select('.chart-tooltip').style('opacity', 0);
}

// Export for use in essay.js
if (typeof window !== 'undefined') {
    window.renderSankey = renderSankey;
}

File: src/essays/charts/sankey.py
"""
Sankey diagram specification for customer flow visualization.

Used in Essay 1: How Customers Actually Become Loyal
Shows the flow of customers between behavior types.
"""

from typing import Any

from src.essays.base import ChartSpec
from src.essays.data_queries import LoyaltyJourneyData


def create_sankey_spec(
    chart_id: str,
    data: LoyaltyJourneyData,
    *,
    width: int = 800,
    height: int = 500,
    node_width: int = 20,
    node_padding: int = 10,
) -> ChartSpec:
    """Create a Sankey diagram specification.

    Args:
        chart_id: Unique identifier for the chart
        data: Loyalty journey data with nodes and links
        width: Chart width in pixels
        height: Chart height in pixels
        node_width: Width of Sankey nodes
        node_padding: Padding between nodes

    Returns:
        ChartSpec for D3 Sankey rendering
    """
    # Build node list with indices and group assignments
    nodes = []
    node_index = {}

    for i, node in enumerate(data.sankey_nodes):
        node_id = node["id"]
        nodes.append({
            "id": node_id,
            "name": node["name"],
            "index": i,
            "group": node.get("group", 0),  # Column position for Sankey layout
        })
        node_index[node_id] = i

    # Build links with source/target indices
    links = []
    for link in data.sankey_links:
        source_id = link["source"]
        target_id = link["target"]

        if source_id in node_index and target_id in node_index:
            links.append({
                "source": node_index[source_id],
                "target": node_index[target_id],
                "value": link["value"],
                "source_id": source_id,
                "target_id": target_id,
            })

    # Scrolly triggers for animation
    scrolly_triggers = [
        {
            "step": 1,
            "action": "highlight_node",
            "params": {"node_id": "new", "description": "All customers start here"},
        },
        {
            "step": 2,
            "action": "highlight_flow",
            "params": {"source": "new", "target": "one_time", "description": "Many drop off immediately"},
        },
        {
            "step": 3,
            "action": "highlight_flow",
            "params": {"source": "new", "target": "regular", "description": "The path to loyalty"},
        },
        {
            "step": 4,
            "action": "show_all",
            "params": {"description": "The complete customer journey"},
        },
    ]

    # Color scheme for behavior types
    colors = {
        "new": "#4299e1",  # Blue
        "one_time": "#fc8181",  # Red
        "irregular": "#f6ad55",  # Orange
        "regular": "#68d391",  # Green
        "long_cycle": "#b794f4",  # Purple
    }

    return ChartSpec(
        chart_id=chart_id,
        chart_type="sankey",
        data={
            "nodes": nodes,
            "links": links,
        },
        config={
            "width": width,
            "height": height,
            "nodeWidth": node_width,
            "nodePadding": node_padding,
            "colors": colors,
            "linkOpacity": 0.5,
            "linkOpacityHover": 0.8,
        },
        scrolly_triggers=scrolly_triggers,
        annotations=[
            {
                "type": "label",
                "node_id": "regular",
                "text": "Champions",
                "position": "right",
            },
        ],
    )


def create_order_distribution_spec(
    chart_id: str,
    data: LoyaltyJourneyData,
    *,
    width: int = 600,
    height: int = 300,
) -> ChartSpec:
    """Create a bar chart spec for order count distribution.

    Args:
        chart_id: Unique identifier
        data: Loyalty journey data
        width: Chart width
        height: Chart height

    Returns:
        ChartSpec for D3 bar chart
    """
    # Convert distribution to bar data
    bars = []
    for order_count, customer_count in sorted(data.order_count_distribution.items()):
        label = f"{order_count}+" if order_count >= 10 else str(order_count)
        bars.append({
            "label": label,
            "value": customer_count,
            "order_count": order_count,
        })

    return ChartSpec(
        chart_id=chart_id,
        chart_type="bar",
        data={"bars": bars},
        config={
            "width": width,
            "height": height,
            "xLabel": "Number of Orders",
            "yLabel": "Customers",
            "color": "#4299e1",
            "highlightColor": "#2b6cb0",
        },
        scrolly_triggers=[
            {
                "step": 1,
                "action": "highlight_bar",
                "params": {"order_count": 1, "description": "One-time buyers"},
            },
            {
                "step": 2,
                "action": "highlight_bar",
                "params": {"order_count": data.loyalty_threshold_order, "description": "Loyalty threshold"},
            },
        ],
    )


def create_retention_line_spec(
    chart_id: str,
    data: LoyaltyJourneyData,
    *,
    width: int = 600,
    height: int = 300,
) -> ChartSpec:
    """Create a line chart spec for retention by order count.

    Args:
        chart_id: Unique identifier
        data: Loyalty journey data
        width: Chart width
        height: Chart height

    Returns:
        ChartSpec for D3 line chart
    """
    # Convert retention data to line points
    points = []
    for order_count, retention in sorted(data.retention_by_order_count.items()):
        points.append({
            "x": order_count,
            "y": retention * 100,  # Convert to percentage
            "label": f"Order {order_count}",
        })

    # Find threshold point for annotation
    threshold = data.loyalty_threshold_order
    threshold_retention = data.retention_by_order_count.get(threshold, 0) * 100

    return ChartSpec(
        chart_id=chart_id,
        chart_type="line",
        data={"points": points},
        config={
            "width": width,
            "height": height,
            "xLabel": "Order Number",
            "yLabel": "Retention Rate (%)",
            "color": "#68d391",
            "strokeWidth": 3,
            "showArea": True,
            "areaOpacity": 0.2,
        },
        scrolly_triggers=[
            {
                "step": 1,
                "action": "draw_line",
                "params": {"duration": 1000},
            },
            {
                "step": 2,
                "action": "add_annotation",
                "params": {
                    "x": threshold,
                    "y": threshold_retention,
                    "text": f"Threshold: Order {threshold}",
                },
            },
        ],
        annotations=[
            {
                "type": "vertical_line",
                "x": threshold,
                "label": f"Loyalty threshold",
                "style": "dashed",
            },
        ],
    )
